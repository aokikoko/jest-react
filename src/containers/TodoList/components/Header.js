import React, { Component } from "react";
// addUndoItem
class Header extends Component {
  constructor(props) {
    super(props);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleInputKeyUp = this.handleInputKeyUp.bind(this);
    this.state = {
      value: "",
    };
  }

  handleInputChange(e) {
    this.setState({ value: e.target.value });
  }
  handleInputKeyUp(e) {
    const { value } = this.state;
    if (e.keyCode === 13 && value) {
      this.props.addUndoItem(value);
      this.setState({ value: "" });
    }
  }
  render() {
    const { value } = this.state;
    return (
      <div className="header">
        <div className="header-content">
          TodoList
          <input
            className="header-input"
            placeholder="Add Todo"
            data-test="input"
            value={value}
            onChange={this.handleInputChange}
            onKeyUp={this.handleInputKeyUp}
          />
        </div>
      </div>
    );
  }
}

export default Header;
