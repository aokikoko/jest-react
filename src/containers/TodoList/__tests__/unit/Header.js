import React from "react";
import { shallow } from "enzyme";
import Header from "../../components/Header";
import { findTestWrapper } from "../../../../utils/testUtils";

describe("Header 组件", () => {
  // 快照
  it("样式渲染正常", () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
  });

  it("组件中包含输入框", () => {
    const wrapper = shallow(<Header />);
    const inputElem = findTestWrapper(wrapper, "input");
    expect(inputElem.length).toBe(1);
  });

  // 受控input框
  it("输入框内容初始为空", () => {
    const wrapper = shallow(<Header />);
    const inputElem = findTestWrapper(wrapper, "input");
    expect(inputElem.prop("value")).toEqual("");
  });

  it("输入框内容随用户输入变化", () => {
    const wrapper = shallow(<Header />);
    const inputElem = findTestWrapper(wrapper, "input");
    const userInput = "啊哈";
    inputElem.simulate("change", {
      target: {
        value: userInput,
      },
    });
    expect(wrapper.state("value")).toEqual(userInput);
    // const newInputElem = wrapper.find("[data-test='input']");
    // expect(newInputElem.prop("value")).toBe(userInput);
  });

  it("输入框无内容时触发回车事件, 无反应", () => {
    const fn = jest.fn();
    const wrapper = shallow(<Header addUndoItem={fn} />);
    const inputElem = findTestWrapper(wrapper, "input");
    wrapper.setState({ value: "" });
    inputElem.simulate("keyup", {
      keyCode: 13,
    });
    expect(fn).not.toHaveBeenCalled();
  });

  it("输入框有内容回车事件被触发时, 外部传入的函数被调用, 内容清空", () => {
    const fn = jest.fn();
    const wrapper = shallow(<Header addUndoItem={fn} />);
    const inputElem = findTestWrapper(wrapper, "input");
    const userInput = "react";
    wrapper.setState({ value: userInput });
    inputElem.simulate("keyup", {
      keyCode: 13,
    });
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenLastCalledWith(userInput);
    const newInputElem = findTestWrapper(wrapper, "input");
    expect(newInputElem.prop("value")).toBe("");
  });
});
