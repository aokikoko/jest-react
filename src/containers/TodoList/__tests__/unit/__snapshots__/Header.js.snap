// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 渲染样式正常 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    TodoList
    <input
      className="header-input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Add Todo"
      value=""
    />
  </div>
</div>
`;
